Create `ProximityAnalysis` model

joshfeli opened the pull request:
This PR features the `ProximityAnalysis` model, which acts as a cache for the results of the analysis functions in `proximity.py`. This PR includes the model, the `admin` registry, a migration file, and a `ProximitySerializer`, which serializes all fields of the model with the caveat of [using the `str` representation](https://www.django-rest-framework.org/api-guide/relations/#stringrelatedfield) of the related `Corpus`. We plan to use the model when implementing the API endpoint to retrieve the results of a proximity analysis. One idea is to have `proximity.run_analysis` return a `ProximityAnalysis` instance by either creating or retrieving one from the database, which an API endpoint could then serialize and ship to frontend.

_Next steps:_ Both `proximity.py` and the model only include functions/fields for the base `run_analysis` method, but the [original file](https://github.com/dhmit/gender_analysis/blob/main/gender_analysis/analysis/proximity.py) from GATK includes many more functions that filter its results. We have yet to transfer those other functions over to the Web project (we've been experimenting on [this branch](https://github.com/dhmit/gender_analysis_web/tree/transfer_proximity)), but fully fleshing out `run_analysis` and its results will lay a strong foundation for when we are ready to include those other functions.

joshfeli committed:
Created a `ProximityAnalysis` model

Commit includes `Model` class, `admin` registry, and a migration file

joshfeli committed:
Created a `ProximitySerializer`

(for the `ProximityAnalysis` model)

MBJean commented on backend/app/serializers.py:
```diff
@@ -62,3 +63,15 @@ class CorpusSerializer(serializers.ModelSerializer):
         fields = ['id', 'title', 'description', 'documents']
+
+
+class ProximitySerializer(serializers.ModelSerializer):
```
Should this be `ProximityAnalysisSerializer`?

joshfeli replied:
Yes! I didn't want the name of the serializer to be too long, but I don't think that will be an issue (seeing as we already have a `PronounSeriesSerializer`). I've renamed it to `ProximityAnalysisSerializer`.

MBJean reviewed (commented):
This looks super straightforward. One thought, especially thinking ahead to when we implement the 'caching' functionality in front of this: should we have some kind of representation of the `Gender` objects used in the analysis? Many a `ManyToMany` relationship? That is: ultimately, the user will be able to select a subset of `Gender`s to include in the analysis, which will be relevant for retrieving the correct `ProximityAnalysis`.

joshfeli committed:
Added a `Gender` `ManyToMany` relation to `ProximityAnalysis`

(Modified migrations accordingly by making a new migration, squashing, and renaming; also specified the `verbose_name_plural` option)

joshfeli committed:
Renamed `ProximitySerializer` to `ProximityAnalysisSerializer`

joshfeli commented:
> This looks super straightforward. One thought, especially thinking ahead to when we implement the 'caching' functionality in front of this: should we have some kind of representation of the `Gender` objects used in the analysis? Many a `ManyToMany` relationship? That is: ultimately, the user will be able to select a subset of `Gender`s to include in the analysis, which will be relevant for retrieving the correct `ProximityAnalysis`.

Makes sense! I've added a `ManyToMany` relation to `Gender`, squashed the migration files, and added the relation to the `ProximityAnalysisSerializer` as a [`PrimaryKeyRelatedField`](https://www.django-rest-framework.org/api-guide/relations/#primarykeyrelatedfield).

joshfeli committed:
Merge branch 'main' into create_proximity_model

MBJean reviewed (approved):
LGTM!

joshfeli committed:
Merged migration files
